lib/autoproj/daemon/buildconf_manager.rb

```diff
@@ -0,0 +1,250 @@
+                    found_branch =
+                        branches.find { |branch| branch.branch_name == branch_name }
+
+                    if found_branch
+                        existing << found_branch
+                    else
+                        Autoproj.message "Creating branch #{branch_name} "\
+                            "on #{buildconf.owner}/#{buildconf.name}"
+                        created << create_branch_for_pr(branch_name, pr)
+                    end
+                end
+
+                [created, existing]
+            end
+
+            OVERRIDES_COMMIT_MSG = 'Update PR overrides'
+            OVERRIDES_FILE = File.join(
+                Autoproj::Workspace::OVERRIDES_DIR,
+                '999-autoproj.yml'
+            ).freeze
+
+            # @param [Github::PullRequest] pull_request
+            # @return [Github::Branch]
+            def create_branch_for_pr(branch_name, pull_request)
+                overrides = overrides_for_pull_request(pull_request)
+                commit_id =
+                    Autoproj::Ops::Snapshot.create_commit(
+                        @main, OVERRIDES_FILE, OVERRIDES_COMMIT_MSG, real_author: false
+                    ) do |io|
+                        YAML.dump(overrides, io)
+                    end
+
+                @importer.run_git_bare(
+                    @main,
+                    'update-ref',
+                    '-m',
+                    OVERRIDES_COMMIT_MSG,
+                    "refs/heads/#{branch_name}",
+                    commit_id
+                )
+                @importer.run_git_bare(
+                    @main, 'push', '-fu', @importer.remote_name, branch_name
+                )
+
+                created = client.branch(buildconf.owner, buildconf.name, branch_name)
+                cache.add(pull_request, overrides)
+                created
+            end
+
+            # @param [Github::PullRequest] pull_request
+            # @return [Array<PackageRepository>]
+            def packages_affected_by_pull_request(pull_request)
+                packages.select do |pkg|
+                    pkg.name == pull_request.base_name &&
+                        pkg.owner == pull_request.base_owner &&
+                        pkg.branch == pull_request.base_branch
+                end
+            end
+
+            # @param [Github::PullRequest] pull_request
+            # @return [Hash]
+            def overrides_for_pull_request(pull_request)
+                # TODO: Extract overrides from PR body.
+                #    This is not as easy as one could have thought
+                #    because the depended upon PR may have dependencies
+                #    of its own. We have to do it recursively. Detecting
+                #    changes in this chain of overrides is even more
+                #    challenging. Or, one could add the entire dependency
+                #    tree in every PR body.
```

Comment:
I am not sure if we should rely on events like that.. Events can be lost, specially if we decide to fetch events for organizations/users (we can can only get the 300 latest events). Also, comments and changes to the PR body are not registered as events. If we want to know once the overrides changes, we have to poll the pull request anyway

Reply:
I've run some tests and I'm pretty sure event polling on a user/organization is not the way to go. Here's a snippet to demonstrate:

```ruby
require 'octokit'

client = Octokit::Client.new
client.auto_paginate = true

event_methods = [
    :user_public_events,
    :user_events,
    :received_public_events,
    :received_events,
    :organization_public_events
]

event_methods.each do |method|
    puts "-- #{method.to_s}"
    client.send(method, 'Microsoft').each do |event|
        next unless event['type'] == 'PullRequestEvent'
        puts "#{event['payload']['pull_request']['base']['repo']['name']}"\
            "##{event['payload']['number']} #{event['payload']['action']}"
    end
    puts
end
```

None of the available methods for event polling on `Microsoft` gives us the PRs on vscode.

Reply:
> Events support pagination, however the per_page option is unsupported. The fixed page size is 30 items. Fetching up to ten pages is supported, for a total of 300 events.

So, not working for initial scan, good to listen to new things.

I have to admit I don't know where the balance between "obviously better technical solution" (query API initially, listen to events afterwards) and "lot less work" (just query API) is.

Since you have the listen-to-events API already, I would say the former won't be a lot more work than the latter - you have to implement querying the API in both cases.

Reply:
It's a bit more work but most of the work is done. My point is that depending on how many events happen during our polling period we may lose events. What I don't know is whether that is acceptable or not. What do you think?

Reply:
> What do you think?

I think it's acceptable, especially looking at the alternative (polling on that many repositories).

Reply:
Fine.. I will mark this as resolved and we get back to the implementation details when I start working on the overrides extraction thing